Running test on a file in the stdlib runs the entire stdlib tests

### Zig Version

0.10.0-dev.4157+4961044ce

### Steps to Reproduce

So, there's multiple things going on here and I'm not sure which of them are bugs, but it seemed like too much information for IRC and I'm pretty sure there's at least one bug in here, so here it is.

### Running `build/stage3/bin/zig test lib/std/fs/path.zig`

Fails to compile with the following error:
```
lib/std/fs/path.zig:2:21: error: import of file outside package path: '../std.zig'
const std = @import("../std.zig");
                    ^~~~~~~~~~~~
referenced by:
    mem: lib/std/fs/path.zig:6:13
    Allocator: lib/std/fs/path.zig:8:19
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
```

This is weird because I remember being told in a Zig community room that this was how you run the tests in the standard library files (that was before stage2/3 was built by default). Is this a breaking change to the way imports work, or a bug?

Same error I pass `-fstage1`. However...

### Running `build/zig2 test lib/std/fs/path.zig`

Builds successfully, but seemingly runs *all* the tests in the standard library, not just those in the named file, which is how I think this used to work.

### Running `build/stage3/bin/zig test lib/std/fs.zig`

Crashes with:
<details>
<summary>expand</summary>

```
Segmentation fault at address 0x7f3438c61d27
???:?:?: 0x8c9b4d4 in getPtrAdapted__anon_219710 (crt/x86_64/crti.s)
???:?:?: 0x8a436b4 in getPtrContext (crt/x86_64/crti.s)
???:?:?: 0x8814f04 in getPtr (crt/x86_64/crti.s)
???:?:?: 0x856a1fa in zirClosureGet (crt/x86_64/crti.s)
???:?:?: 0x83c116e in analyzeBodyInner (crt/x86_64/crti.s)
???:?:?: 0x83bf608 in analyzeBody (crt/x86_64/crti.s)
???:?:?: 0x881261c in analyzeCall (crt/x86_64/crti.s)
???:?:?: 0x8569eff in zirCall (crt/x86_64/crti.s)
???:?:?: 0x83c1090 in analyzeBodyInner (crt/x86_64/crti.s)
???:?:?: 0x81fd698 in analyzeBodyBreak (crt/x86_64/crti.s)
???:?:?: 0x83d051d in analyzeBodyInner (crt/x86_64/crti.s)
???:?:?: 0x83bf608 in analyzeBody (crt/x86_64/crti.s)
???:?:?: 0x881261c in analyzeCall (crt/x86_64/crti.s)
???:?:?: 0x8569eff in zirCall (crt/x86_64/crti.s)
???:?:?: 0x83c1090 in analyzeBodyInner (crt/x86_64/crti.s)
???:?:?: 0x81fd698 in analyzeBodyBreak (crt/x86_64/crti.s)
???:?:?: 0x83d051d in analyzeBodyInner (crt/x86_64/crti.s)
???:?:?: 0x83bf608 in analyzeBody (crt/x86_64/crti.s)
???:?:?: 0x881261c in analyzeCall (crt/x86_64/crti.s)
???:?:?: 0x8569eff in zirCall (crt/x86_64/crti.s)
???:?:?: 0x83c1090 in analyzeBodyInner (crt/x86_64/crti.s)
???:?:?: 0x81fd698 in analyzeBodyBreak (crt/x86_64/crti.s)
???:?:?: 0x880c5c5 in resolveBody (crt/x86_64/crti.s)
???:?:?: 0x881fd5f in resolveGenericBody (crt/x86_64/crti.s)
???:?:?: 0x8573fc3 in zirFunc (crt/x86_64/crti.s)
???:?:?: 0x83c2db7 in analyzeBodyInner (crt/x86_64/crti.s)
???:?:?: 0x81fd698 in analyzeBodyBreak (crt/x86_64/crti.s)
???:?:?: 0x880c5c5 in resolveBody (crt/x86_64/crti.s)
???:?:?: 0x8a3ea4b in instantiateGenericCall (crt/x86_64/crti.s)
???:?:?: 0x880e9a4 in analyzeCall (crt/x86_64/crti.s)
???:?:?: 0x8569eff in zirCall (crt/x86_64/crti.s)
???:?:?: 0x83c1090 in analyzeBodyInner (crt/x86_64/crti.s)
???:?:?: 0x81fd698 in analyzeBodyBreak (crt/x86_64/crti.s)
???:?:?: 0x880c5c5 in resolveBody (crt/x86_64/crti.s)
???:?:?: 0x8569bd3 in zirCall (crt/x86_64/crti.s)
???:?:?: 0x83c1090 in analyzeBodyInner (crt/x86_64/crti.s)
???:?:?: 0x83bf608 in analyzeBody (crt/x86_64/crti.s)
???:?:?: 0x81e5d25 in analyzeFnBody (crt/x86_64/crti.s)
???:?:?: 0x7fcf288 in ensureFuncBodyAnalyzed (crt/x86_64/crti.s)
???:?:?: 0x7fcd753 in processOneJob (crt/x86_64/crti.s)
???:?:?: 0x7eac40a in performAllTheWork (crt/x86_64/crti.s)
???:?:?: 0x7ea8c4b in update (crt/x86_64/crti.s)
???:?:?: 0x7ed05c6 in updateModule (crt/x86_64/crti.s)
???:?:?: 0x7c7433c in buildOutputType (crt/x86_64/crti.s)
???:?:?: 0x7c52369 in mainArgs (crt/x86_64/crti.s)
???:?:?: 0x7f39c77 in main (crt/x86_64/crti.s)
fish: Job 1, 'build/stage3/bin/zig test lib...' terminated by signal SIGABRT (Abort)
```

</details>

### Running `build/stage3/bin/zig test -fstage1 lib/std/fs.zig`

Fails to compile with:
<details>
<Summary>expand</summary>

```
./lib/std/meta.zig:471:29: error: expected type '[]const builtin.Type.EnumField', found '[]const std.builtin.Type.EnumField'
        .Enum => |info| info.fields,
                            ^
./lib/std/enums.zig:735:41: note: called from here
    const const_fields = std.meta.fields(E);
                                        ^
./lib/std/enums.zig:839:32: note: called from here
    const Indexer = EnumIndexer(E);
                               ^
./lib/std/enums.zig:837:37: note: called from here
test "std.enums.EnumIndexer sparse" {
                                    ^
./lib/std/meta.zig:471:29: note: slice type child 'std.builtin.Type.EnumField' cannot cast into slice type child 'builtin.Type.EnumField'
        .Enum => |info| info.fields,
                            ^
./lib/std/builtin.zig:327:27: note: builtin.Type.EnumField declared here
    pub const EnumField = struct {
                          ^
./build/stage3/lib/zig/std/builtin.zig:327:27: note: std.builtin.Type.EnumField declared here
    pub const EnumField = struct {
```

</details>

The full error output is too long to include; it complains about errors like this - `std.builtin` where `builtin` was expected, or vice versa - everywhere.

This also looks like a symptom of a breaking change, but I *just* built this from master, so it can't be that my compiler build is out of sync with the test sources.


### Expected Behavior

Any of these commands should run only the tests in the named files. This is how I remember it working before.

### Actual Behavior

See above

--- a/src/Autodoc.zig
+++ b/src/Autodoc.zig
@@ -853,7 +853,7 @@ fn walkInstruction(
             var path = str_tok.get(file.zir);
 
             const maybe_other_package: ?*Package = blk: {
-                if (self.module.main_pkg_in_std and std.mem.eql(u8, path, "std")) {
+                if (self.module.main_pkg_is_std and std.mem.eql(u8, path, "std")) {
                     path = "root";
                     break :blk self.module.main_pkg;
                 } else {

--- a/src/Compilation.zig
+++ b/src/Compilation.zig
@@ -1588,10 +1588,10 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
             try main_pkg.add(gpa, "root", root_pkg);
             try main_pkg.addAndAdopt(gpa, "std", std_pkg);
 
-            const main_pkg_in_std = m: {
+            const main_pkg_is_std = m: {
                 const std_path = try std.fs.path.resolve(arena, &[_][]const u8{
                     std_pkg.root_src_directory.path orelse ".",
-                    std.fs.path.dirname(std_pkg.root_src_path) orelse ".",
+                    std_pkg.root_src_path,
                 });
                 defer arena.free(std_path);
                 const main_path = try std.fs.path.resolve(arena, &[_][]const u8{
@@ -1599,7 +1599,7 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
                     main_pkg.root_src_path,
                 });
                 defer arena.free(main_path);
-                break :m mem.startsWith(u8, main_path, std_path);
+                break :m mem.eql(u8, main_path, std_path);
             };
 
             // Pre-open the directory handles for cached ZIR code so that it does not need
@@ -1638,7 +1638,7 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
                 .gpa = gpa,
                 .comp = comp,
                 .main_pkg = main_pkg,
-                .main_pkg_in_std = main_pkg_in_std,
+                .main_pkg_is_std = main_pkg_is_std,
                 .root_pkg = root_pkg,
                 .zig_cache_artifact_directory = zig_cache_artifact_directory,
                 .global_zir_cache = global_zir_cache,

--- a/src/Module.zig
+++ b/src/Module.zig
@@ -142,7 +142,7 @@ job_queued_update_builtin_zig: bool = true,
 /// This makes it so that we can run `zig test` on the standard library.
 /// Otherwise, the logic for scanning test decls skips all of them because
 /// `main_pkg != std_pkg`.
-main_pkg_in_std: bool,
+main_pkg_is_std: bool,
 
 compile_log_text: ArrayListUnmanaged(u8) = .{},
 
@@ -5171,7 +5171,7 @@ fn scanDecl(iter: *ScanDeclIter, decl_sub_index: usize, flags: u4) Allocator.Err
                 // the test name filter.
                 if (!comp.bin_file.options.is_test) break :blk false;
                 if (decl_pkg != mod.main_pkg) {
-                    if (!mod.main_pkg_in_std) break :blk false;
+                    if (!mod.main_pkg_is_std) break :blk false;
                     const std_pkg = mod.main_pkg.table.get("std").?;
                     if (std_pkg != decl_pkg) break :blk false;
                 }
@@ -5182,7 +5182,7 @@ fn scanDecl(iter: *ScanDeclIter, decl_sub_index: usize, flags: u4) Allocator.Err
                 if (!is_named_test) break :blk false;
                 if (!comp.bin_file.options.is_test) break :blk false;
                 if (decl_pkg != mod.main_pkg) {
-                    if (!mod.main_pkg_in_std) break :blk false;
+                    if (!mod.main_pkg_is_std) break :blk false;
                     const std_pkg = mod.main_pkg.table.get("std").?;
                     if (std_pkg != decl_pkg) break :blk false;
                 }

